src/radical/pilot/agent/agent_0.py

```diff
@@ -354,106 +356,137 @@ def _start_services(self):
+            #                         'core_map' : [[0]],
+            #                         'gpu_map'  : [],
+            #                         'lfs'      : 0,
+            #                         'mem'      : 0}]}
+        # }
+
+        # launcher = self._rm.find_launcher(service_task)
+        # if not launcher:
+        #     raise RuntimeError('no launch method found for sub agent')
 
-        with ru.ru_open(ex_name, 'w') as fout:
-            fout.write(tmp)
+        # FIXME: set RP environment (as in Popen Executor)
 
+        # tmp  = '#!/bin/sh\n\n'
+        # tmp += 'export RP_PILOT_SANDBOX="%s"\n\n' % self._pwd
+        # cmds = launcher.get_launcher_env()
+        # for cmd in cmds:
+        #     tmp += '%s || exit 1\n' % cmd
+        # cmds = launcher.get_launch_cmd(service_task, ex_name)
+        # tmp += '%s\nexit $?\n\n' % '\n'.join(cmds)
+        #
+        # with ru.ru_open(ls_name, 'w') as fout:
+        #     fout.write(tmp)
 
+        # tmp  = '#!/bin/sh\n\n'
+        # tmp += '. ./env/service.env\n'
+        # tmp += '/bin/sh -l ./services\n\n'
+        #
+        # with ru.ru_open(ex_name, 'w') as fout:
+        #     fout.write(tmp)
+        #
+        #
         # make sure scripts are executable
-        st_l = os.stat(ls_name)
-        st_e = os.stat(ex_name)
-        os.chmod(ls_name, st_l.st_mode | stat.S_IEXEC)
-        os.chmod(ex_name, st_e.st_mode | stat.S_IEXEC)
-
+        # st_l = os.stat(ls_name)
+        # st_e = os.stat(ex_name)
+        # os.chmod(ls_name, st_l.st_mode | stat.S_IEXEC)
+        # os.chmod(ex_name, st_e.st_mode | stat.S_IEXEC)
+        #
         # spawn the sub-agent
-        cmdline = './%s' % ls_name
+        # cmdline = './%s' % ls_name
+        #
+        # self._log.info('create services: %s' % cmdline)
+        # ru.sh_callout_bg(cmdline, stdout='services.out', stderr='services.err')
 
-        self._log.info('create services: %s' % cmdline)
-        ru.sh_callout_bg(cmdline, stdout='services.out', stderr='services.err')
-        e = threading.Event()
-        self.publish(rpc.STATE_PUBSUB, {'cmd': 'service_state_update','arg': {services , e}})
-        e.set()
-        self._log.debug('services started done')
+        # self.publish(rpc.STATE_PUBSUB, {'cmd': 'service_state_update','arg': services})
 
     def _state_cb_of_services(self, topic, msg):
         cmd = msg['cmd']
-        arg, e = msg['arg']
-        if cmd == 'service_state_update':
+        services = msg['arg']
 
-            tasks = ru.as_list(arg)
+        if cmd == 'update':
+            self._log.info('received services for launching: %s' % services)
+            # e = threading.Event()
+            e.set()
 
-            for task in tasks:
-                uid   = task['uid']
-                state = task['state']
+            self._log.debug('services started done')
 
-                if state == rps.AGENT_EXECUTING:
+            running_services = []
+            yet_to_run = []
```

Comment:
I don't think you need to track services which are not in the `AGENT_EXECUTING` state.  Also, you will double-count because the service tasks go through several states before `AGENT_EXECUTING`.

Reply:
Ohh. Yes. The `yet_to_run` can get multiple entries for different states of same task. Ofcourse we can solve that easily, but I am not sure of what I'll do with this `yet_to_run` in future. So will remove